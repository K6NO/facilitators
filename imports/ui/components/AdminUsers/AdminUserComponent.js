/* eslint-disable max-len, no-return-assign */
import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'reactstrap';
import AdminSelectOrganisationComponent from './AdminSelectOrganisationComponent';
import AdminUserRoleSelector from './AdminUserRoleSelector';
import DeleteUserButton from './DeleteUserButton';
import InactivateUserButton from './InactivateUserButton';
import saveUserProfile from './saveUserProfile';
import storePassword from './storePassword';
import './AdminUserComponent.scss';

class AdminUserComponent extends React.Component {
  constructor(props){
    super(props);
    this.state = {
        user : {
          name: props.userListed.profile && 
          props.userListed.profile.name.first + ' ' 
          + props.userListed.profile.name.last || '',
          email: props.userListed.emails && 
          props.userListed.emails[0].address || '',
          password : '',
          username: props.userListed.profile && 
          props.userListed.profile.username || '',
          organisation: props.userListed.organisation || '',
          roles: props.userListed.roles || [],
        },
        editing: false,
        passwordEditing : false,
      };
  }

  // Callbacks setting the state
  updateUserState = (field, event) => {
    this.setState({
      user : {
        ...this.state.user,
        [field]: event.target.value,
      }
    });
  }

  selectOrganisationCallback = (_id) => {
    this.setState({
        user : {
          ...this.state.user,
          organisation: _id,
        }
      });                       
  }
  switchUserRolesCallback = (roles) => {
        this.setState({
        ...this.state,
        user: {
            ...this.state.user,
            roles: roles,
        }
    });
  }
  resetUserStateCallback = () => {
    this.setState({
        user : {
              name: '',
              password: '',
              email: '',
              username: '',
              organisation: '',
              roles: [],
          },
      });
  }
  // Switch the main editing element's state
  switchEditing = () => {
    if(this.state.editing) {
        const { userListed, closeCallback } = this.props;
        saveUserProfile(userListed, this.state.user, closeCallback, this.resetUserStateCallback);
    }
    this.setState({
        ...this.state,
        editing: !this.state.editing,
    });
  }
  // switch the password editing element's state
  switchPasswordEditing = () => {
    const { userListed } = this.props;
    if(this.state.passwordEditing) {
        const isNewUser = storePassword(userListed, this.state.user);
        // Reset state for new and existing user when edit mode 
        isNewUser ? (
            this.setState({
                user: {
                    ...this.state.user,
                    password: '',
                },
                passwordEditing: !this.state.passwordEditing,
            })) : (
                this.setState({
                    passwordEditing: !this.state.passwordEditing,
                }));
    } else {
        // switch on edit mode
        this.setState({
            passwordEditing: !this.state.passwordEditing,
        });
    }
  }

  render() {
    const { organisations, userListed } = this.props;
    return (
        <tr className="AdminUserComponent">
            <td>
                {this.state.editing ? (
                    <input 
                        name="name" 
                        type="text"
                        className="formControl"
                        onChange={(e) => this.updateUserState('name', e)} 
                        value={this.state.user.name} />
                ) : (
                    <span>{this.state.user.name}</span> 
                )}
            </td>
            <td>
                {this.state.editing ? (
                <input 
                    name="email" 
                    type="text" 
                    className="formControl"
                    onChange={(e) => this.updateUserState('email', e)} 
                    value={this.state.user.email} />
                ) : (
                    <span>{this.state.user.email}</span> 
                )}
            </td>
            <td>
                {this.state.editing ? (
                    this.state.passwordEditing ? (
                        <div>
                            <input 
                                name="password" 
                                type="text" 
                                className="formControl"
                                onChange={(e) => this.updateUserState('password', e)} 
                                value={this.state.user.password}
                                placeholder="New password" />
                            <Button 
                                className="savePasswordButton"
                                onClick={this.switchPasswordEditing}>Save password</Button>
                        </div>
                    ) : 
                    <Button
                        className="switchPasswordButton"
                        onClick={this.switchPasswordEditing}>
                        Reset password</Button>
                ) : ''}
            </td>
            <td>
                {this.state.editing ? (
                    <input 
                        name="username" 
                        type="text" 
                        className="formControl"
                        onChange={(e) => this.updateUserState('username', e)} 
                        value={this.state.user.username} />
                ) : (
                    <span>{this.state.user.username}</span> 
                )}
            </td>
            <td>
                 {organisations.length > 0 ? 
                    <AdminSelectOrganisationComponent 
                    editing={this.state.editing}
                    organisations={organisations}
                    selectOrganisationCallback={this.selectOrganisationCallback}
                    selectedOrganisation={this.state.user.organisation}
                    user={this.props.userListed}
                /> 
                : 'No organisations.'}
            </td>
            <td>
                <AdminUserRoleSelector 
                    editing={this.state.editing}
                    user={this.props.userListed}
                    switchUserRolesCallback={this.switchUserRolesCallback}
                />
            </td>
            <td>
                {Roles.userIsInRole(userListed._id, ['admin'])
                    ? '' 
                    :
                    <DeleteUserButton 
                        userId={this.props.userListed._id}/>
                }
            </td>
            <td>
                {Roles.userIsInRole(userListed._id, ['admin'])
                    ? '' 
                    :
                    <InactivateUserButton 
                        userListed={this.props.userListed}/>
                }
            </td>
            <td>
                <Button
                    className="switchEditModeButton"
                    onClick={this.switchEditing}>
                    {this.state.editing ? 'Save' : 'Edit'}
                </Button>
            </td>
        </tr>
    );
  }
}

AdminUserComponent.defaultProps = {
  userListed : {},
};

AdminUserComponent.propTypes = {
  userListed: PropTypes.object.isRequired,
  organisations: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default AdminUserComponent;