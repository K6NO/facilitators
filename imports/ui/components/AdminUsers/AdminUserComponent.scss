.AdminUserComponent {
    .formControl {
        height: 2.5rem;
        border-radius: 0;
    }
}