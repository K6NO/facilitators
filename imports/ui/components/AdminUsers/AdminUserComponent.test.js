import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import AdminUserComponent from './AdminUserComponent';
import { Button } from 'reactstrap';



const user = {
    _id: "UserId1",
    roles: ['admin', 'editor', 'superadmin']
};
const organisations = [{
    _id: '123',
    name: 'org1name',
    }, {
        _id: '456',
        name: 'org2name',
    }
];
test('AdminUserComponent renders with props', () => {
    const component = renderer.create(
        <AdminUserComponent 
            key={user._id} 
            userListed={user}
            organisations={organisations}
        />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Clicking SwitchPasswordButton opens password editing mode', () => {
    const shallowComp = shallow(
        <AdminUserComponent 
            key={user._id} 
            userListed={user}
            organisations={organisations}
        />);
        shallowComp.setState({editing:true});
        shallowComp.find('.switchPasswordButton').simulate('click');
        // to get refreshed state
        shallowComp.update();
        expect(shallowComp.state().passwordEditing).toBeTruthy();
});

test('Clicking Edit button opens general editing mode', () => {
    const shallowComp = shallow(
        <AdminUserComponent 
            key={user._id} 
            userListed={user}
            organisations={organisations}
        />);
        shallowComp.find('.switchEditModeButton').simulate('click');
        // to get refreshed state
        shallowComp.update();
        expect(shallowComp.state().editing).toBeTruthy();
});