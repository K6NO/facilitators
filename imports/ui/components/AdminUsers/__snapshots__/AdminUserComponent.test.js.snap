// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminUserComponent renders with props 1`] = `
<tr>
  <td>
    <span>
      
    </span>
  </td>
  <td>
    <span>
      
    </span>
  </td>
  <td>
    
  </td>
  <td>
    <span>
      
    </span>
  </td>
  <td>
    <span>
      
    </span>
  </td>
  <td>
    <span>
      admin, editor, superadmin, 
    </span>
  </td>
  <td>
    
  </td>
  <td>
    <button
      aria-label={null}
      className="switchEditModeButton btn btn-secondary"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </td>
</tr>
`;
